// react-native-simple-widgets/dist/commonjs/MaskInput/createNumberMask.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.createNumberMaskByLocale = void 0;

var _currencyformat = require("../utils/currencyformat");

function createNumberMask(props) {
  const {
    delimiter = ".",
    precision = 2,
    prefix = ["R", "$", " "],
    separator = ","
  } = props || {};
  return value => {
    const numericValue = (value === null || value === void 0 ? void 0 : value.replace(/\D+/g, "")) || "";
    const mask = numericValue.split("").map(() => /\d/);

    if (precision > 0 && mask.length > precision) {
      mask.splice(-precision, 0, separator);
    }

    const delimiters = Math.ceil((numericValue.length - precision) / 3) - 1;

    for (let i = 1; i <= delimiters; i++) {
      const position = -precision - i * 3 - (precision > 0 ? i : i - 1);
      mask.splice(position, 0, delimiter);
    }

    return [...prefix, ...mask];
  };
}

const createNumberMaskByLocale = locale => {
  const decimalSeparator = (0, _currencyformat.localeCurrencyDecimalSeparator)(locale);
  return createNumberMask({
    prefix: [],
    separator: decimalSeparator,
    delimiter: decimalSeparator === "," ? "." : ",",
    precision: decimalSeparator === "," ? 0 : 2
  });
};

exports.createNumberMaskByLocale = createNumberMaskByLocale;
var _default = createNumberMask;
exports.default = _default;
