// react-native-simple-widgets/dist/commonjs/MultipleSlider/DefaultMarker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class DefaultMarker extends _react.default.Component {
  render() {
    return /*#__PURE__*/_react.default.createElement(_reactNative.TouchableHighlight, null, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: this.props.enabled ? [styles.markerStyle, this.props.markerStyle, this.props.pressed && styles.pressedMarkerStyle, this.props.pressed && this.props.pressedMarkerStyle] : [styles.markerStyle, styles.disabled, this.props.disabledMarkerStyle]
    }));
  }

}

const styles = _reactNative.StyleSheet.create({
  markerStyle: { ..._reactNative.Platform.select({
      ios: {
        height: 30,
        width: 30,
        borderRadius: 30,
        borderWidth: 1,
        borderColor: "#DDDDDD",
        backgroundColor: "#FFFFFF",
        shadowColor: "#000000",
        shadowOffset: {
          width: 0,
          height: 3
        },
        shadowRadius: 1,
        shadowOpacity: 0.2
      },
      android: {
        height: 12,
        width: 12,
        borderRadius: 12,
        backgroundColor: "#0D8675"
      },
      web: {
        height: 30,
        width: 30,
        borderRadius: 30,
        borderWidth: 1,
        borderColor: "#DDDDDD",
        backgroundColor: "#FFFFFF",
        shadowColor: "#000000",
        shadowOffset: {
          width: 0,
          height: 3
        },
        shadowRadius: 1,
        shadowOpacity: 0.2
      }
    })
  },
  pressedMarkerStyle: { ..._reactNative.Platform.select({
      web: {},
      ios: {},
      android: {
        height: 20,
        width: 20,
        borderRadius: 20
      }
    })
  },
  disabled: {
    backgroundColor: "#d3d3d3"
  }
});

var _default = DefaultMarker;
exports.default = _default;
