// react-native-simple-widgets/dist/commonjs/AuthorizedUserMenu/AuthorizedUserMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _Avatar = require("react-native-simple-elements/components/Avatar");

var _Menu = _interopRequireDefault(require("react-native-simple-elements/components/Menu"));

var _ImageButton = _interopRequireDefault(require("react-native-simple-elements/components/ImageButton"));

var _Divider = _interopRequireDefault(require("react-native-simple-elements/components/Divider"));

var _TouchableRipple = _interopRequireDefault(require("react-native-simple-elements/components/TouchableRipple"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const parseAvatarText = fullName => {
  const parts = (fullName || "").split(" ");
  let text = "";

  if (parts.length >= 1) {
    text = text + parts[0].charAt(0).toUpperCase();
  }

  if (parts.length >= 2) {
    text = text + parts[1].charAt(0).toUpperCase();
  }

  return text;
};

const defaultProps = {
  circle: false
};

const AuthorizedUserMenu = ({
  loggedInUser,
  circle,
  avatarSize,
  avatarStyle,
  avatarLabelStyle,
  onViewProfileClick,
  onLogoutClick,
  userMenuItems,
  onUserMenuItemPress
}) => {
  const {
    firstName,
    avatarUrl
  } = loggedInUser || {};
  const [isOpen, setIsOpen] = React.useState(false);

  const openMenu = () => {
    setIsOpen(true);
  };

  const closeMenu = () => {
    setIsOpen(false);
  };

  const _handleMenuItemPress = item => {
    if (onUserMenuItemPress) {
      onUserMenuItemPress(item);
    }
  };

  return /*#__PURE__*/React.createElement(_Menu.default, {
    visible: isOpen,
    onDismiss: closeMenu,
    anchor: loggedInUser !== null && loggedInUser !== void 0 && loggedInUser.userId && avatarUrl ? /*#__PURE__*/React.createElement(_ImageButton.default, {
      source: {
        uri: avatarUrl
      },
      circle: circle,
      onPress: openMenu
    }) : /*#__PURE__*/React.createElement(_TouchableRipple.default, {
      onPress: openMenu
    }, /*#__PURE__*/React.createElement(_Avatar.AvatarText, {
      label: parseAvatarText(loggedInUser === null || loggedInUser === void 0 ? void 0 : loggedInUser.fullName),
      size: avatarSize,
      style: avatarStyle,
      labelStyle: avatarLabelStyle
    })),
    alignRight: true
  }, (loggedInUser === null || loggedInUser === void 0 ? void 0 : loggedInUser.userId) && /*#__PURE__*/React.createElement(_Menu.default.Item, {
    title: firstName || "",
    titleStyle: {
      textAlign: "center"
    },
    onPress: () => {
      if (typeof onViewProfileClick === "function") {
        onViewProfileClick();
      }
    }
  }), /*#__PURE__*/React.createElement(_Divider.default, null), Array.isArray(userMenuItems) && userMenuItems.length > 0 ? userMenuItems.map((item, index) => {
    return /*#__PURE__*/React.createElement(_Menu.default.Item, {
      key: index,
      title: item.label,
      icon: item.icon,
      onPress: () => _handleMenuItemPress(item)
    });
  }) : null, /*#__PURE__*/React.createElement(_Divider.default, null), (loggedInUser === null || loggedInUser === void 0 ? void 0 : loggedInUser.userId) && /*#__PURE__*/React.createElement(_Menu.default.Item, {
    title: "Logout",
    onPress: () => {
      if (typeof onLogoutClick === "function") {
        onLogoutClick();
      }
    }
  }));
};

AuthorizedUserMenu.defaultProps = defaultProps;
var _default = AuthorizedUserMenu;
exports.default = _default;
