// react-native-simple-elements/dist/module/Toast/Toast/index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import * as React from "react";
import { Animated, TouchableOpacity, Vibration } from "react-native";
import { getStatusBarHeight } from "react-native-iphone-x-helper";
import CheckCircleIcon from "@mdi/svg/svg/account.svg";
import AlertCircleIcon from "@mdi/svg/svg/account.svg";
import XCircleIcon from "@mdi/svg/svg/account.svg";
import CloseIcon from "@mdi/svg/svg/account.svg";
import Box from "../Box";
import { SvgIcon } from "../../Icon";
import { Accent, Heading, IconCont, StyledToast, SubText } from "./styles"; // type IconFamilies =
//     | 'Entypo'
//     | 'EvilIcons'
//     | 'Feather'
//     | 'FontAwesome'
//     | 'Foundation'
//     | 'Ionicons'
//     | 'MaterialCommunityIcons'
//     | 'MaterialIcons'
//     | 'Octicons'
//     | 'SimpleLineIcons'
//     | 'Zocial'

const statusBarHeight = getStatusBarHeight();
const offset = statusBarHeight + 16;
const shadow = {
  shadowColor: "#000",
  shadowOffset: {
    width: 0,
    height: 0
  },
  shadowOpacity: 0.1,
  shadowRadius: 2,
  elevation: 1
};
const DEFAULT_PROPS = {
  duration: 3000,
  intent: "SUCCESS",
  onPress: () => false,
  shouldVibrate: false,
  closeIconColor: "text",
  message: "Toast message!",
  hideIcon: false,
  toastStyles: {
    borderColor: "black",
    bg: "background"
  },
  closeButtonStyles: {
    p: 2,
    mx: 2,
    bg: "black",
    borderRadius: 4,
    alignItems: "center"
  }
};
export const Toast = ({
  accentColor,
  closeIconColor,
  // closeIconFamily,
  // closeIconName,
  closeIconSize,
  color,
  duration,
  hideIcon,
  iconColor,
  // iconFamily,
  iconName,
  id,
  index,
  intent,
  message,
  onClose,
  onPress,
  position,
  shouldVibrate,
  subMessage,
  toastStyles,
  hideAccent,
  closeButtonStyles
}) => {
  const isSuccess = intent === "SUCCESS";
  const isInfo = intent === "INFO";
  const topOffset = offset + 60 * (index || 0);
  const animation = React.useRef(new Animated.Value(0)).current;
  React.useEffect(() => {
    Animated.timing(animation, {
      toValue: 0.5,
      duration: 300,
      useNativeDriver: true
    }).start(() => {
      if (duration !== 0) {
        const timer = setTimeout(() => {
          if (index === 0) {
            clearTimeout(timer);
          }

          id && onClose && onClose(id);
        }, duration);
      }
    });

    if (shouldVibrate) {
      Vibration.vibrate(10);
    }
  }, []);
  const translateY = animation.interpolate({
    inputRange: [0, 0.5],
    outputRange: [position === "BOTTOM" ? topOffset : -topOffset, 0]
  });
  const scale = animation.interpolate({
    inputRange: [0, 0.5],
    outputRange: [0.8, 1],
    extrapolate: "clamp"
  });
  return /*#__PURE__*/React.createElement(StyledToast, _extends({
    onPress: () => {
      onPress && onPress();
      onClose && id && onClose(id);
    },
    style: {
      transform: [{
        translateY
      }, {
        scale
      }],
      ...shadow
    }
  }, toastStyles, {
    pr: subMessage ? 2 : 0
  }), !hideAccent && /*#__PURE__*/React.createElement(Accent, {
    testID: "toast-accent",
    bg: accentColor ? accentColor : isSuccess ? "success" : isInfo ? "info" : "error"
  }), !hideIcon && /*#__PURE__*/React.createElement(IconCont, {
    px: 4
  }, /*#__PURE__*/React.createElement(SvgIcon, {
    icon: iconName ? iconName : isSuccess ? CheckCircleIcon : isInfo ? AlertCircleIcon : XCircleIcon,
    size: 20,
    color: iconColor ? iconColor : isSuccess ? "success" : isInfo ? "info" : "error" // family={iconFamily || 'Feather'}
    // name={!!iconName ? iconName : isSuccess ? CheckCircleIcon : isInfo ? AlertCircleIcon : XCircleIcon}

  })), /*#__PURE__*/React.createElement(Box, {
    alignItems: "flex-start",
    flex: "1 1 auto",
    pl: hideIcon ? 4 : 0,
    pr: subMessage ? 2 : 0,
    py: 2
  }, /*#__PURE__*/React.createElement(Box, {
    flexDirection: "row",
    flexWrap: "wrap",
    flex: "1 1 auto"
  }, /*#__PURE__*/React.createElement(Heading, {
    color: color
  }, message)), !!subMessage && /*#__PURE__*/React.createElement(SubText, {
    color: color,
    mt: 1
  }, subMessage)), /*#__PURE__*/React.createElement(TouchableOpacity, {
    onPress: () => onClose && id && onClose(id)
  }, /*#__PURE__*/React.createElement(Box, Object.assign({}, DEFAULT_PROPS.closeButtonStyles, closeButtonStyles), /*#__PURE__*/React.createElement(SvgIcon, {
    icon: CloseIcon,
    size: closeIconSize || 20,
    color: closeIconColor // family={closeIconFamily || 'Feather'}
    // name={closeIconName || 'x'}

  }))));
};
export default /*#__PURE__*/React.memo(Toast);
Toast.defaultProps = DEFAULT_PROPS;
