// react-native-simple-elements/dist/module/Toast/Toast/styles.js
function _templateObject5() {
  const data = _taggedTemplateLiteral(["\n  ", ";\n  align-items: center;\n  justify-content: center;\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  const data = _taggedTemplateLiteral(["\n  ", ";\n  align-items: center;\n  justify-content: center;\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  const data = _taggedTemplateLiteral(["\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  flex-wrap: wrap;\n  font-weight: normal;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  const data = _taggedTemplateLiteral(["\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  flex-wrap: wrap;\n  font-weight: normal;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  const data = _taggedTemplateLiteral(["\n  ", ";\n  top: 0;\n  left: 0;\n  bottom: 0;\n  width: 4px;\n  position: absolute;\n  border-top-left-radius: 4px;\n  border-bottom-left-radius: 4px;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import { Animated } from "react-native";
import styled from "styled-components/native";
import { background, border, color, compose, flexbox, fontSize, layout, position, shadow, space, textAlign } from "styled-system";
export const systemProps = compose(space, layout, color, flexbox, background, border, position, shadow);
// const AnimatedTouchable = Animated.createAnimatedComponent(TouchableOpacity);
export const StyledToast = Animated.createAnimatedComponent(styled.TouchableOpacity(systemProps));
StyledToast.defaultProps = {
  py: 2,
  mb: 4,
  flex: "1 1 auto",
  zIndex: 1000,
  width: "100%",
  bg: "background",
  borderWidth: "1px",
  borderRadius: "4px",
  alignItems: "center",
  flexDirection: "row",
  accentColor: "green",
  justifyContent: "center",
  borderColor: "black"
};
export const Accent = styled.View(_templateObject(), color);
export const Heading = styled.Text(_templateObject2(), space, color, fontSize, textAlign);
Heading.defaultProps = {
  fontSize: 2,
  color: "text"
};
export const SubText = styled.Text(_templateObject3(), space, color, fontSize, textAlign);
SubText.defaultProps = {
  fontSize: 1,
  color: "text"
};
export const IconCont = styled.View(_templateObject4(), space);
export const CloseButtonCont = styled.TouchableOpacity(_templateObject5(), space);
