// react-native-simple-widgets/dist/commonjs/ViewportVisibilitySensor/ViewportVisibilitySensor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ViewportVisibilitySensor extends React.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "state", {
      prevRectTop: 0,
      rectTop: 0,
      rectBottom: 0
    });

    _defineProperty(this, "interval", void 0);

    _defineProperty(this, "myview", void 0);

    _defineProperty(this, "lastValue", void 0);
  }

  componentDidMount() {
    if (!this.props.disabled) {
      this.startWatching();
    }
  }

  componentWillUnmount() {
    this.stopWatching();
  }

  startWatching() {
    if (this.interval) {
      return;
    }

    this.interval = setInterval(() => {
      if (!this.myview) {
        return;
      }

      this.myview.measure((x, y, width, height, pageX, pageY) => {
        this.setState({
          prevRectTop: this.state.rectTop,
          rectTop: pageY,
          rectBottom: pageY + height,
          rectWidth: pageX + width
        });
        this.isInViewPort();
      });
    }, this.props.delay || 100);
  }

  stopWatching() {
    this.interval = clearInterval(this.interval);
  }

  isInViewPort() {
    const window = _reactNative.Dimensions.get("window");

    const isVisible = this.state.rectBottom !== 0 && this.state.rectTop >= 0 && this.state.rectBottom <= window.height && this.state.rectWidth > 0 && this.state.rectWidth <= window.width;

    if (this.lastValue !== isVisible) {
      this.lastValue = isVisible; // console.log("Top ", this.state.rectTop, " Prev top ", this.state.prevRectTop);

      this.props.onChange(isVisible, {
        direction: this.state.rectTop > this.state.prevRectTop ? -1 : 1
      });
    }
  }

  render() {
    return /*#__PURE__*/React.createElement(_reactNative.View, _extends({
      collapsable: false,
      ref: component => {
        this.myview = component;
      }
    }, this.props), this.props.children);
  }

}

var _default = ViewportVisibilitySensor;
exports.default = _default;
