// react-native-simple-elements/dist/commonjs/Drawer/DrawerItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _color = _interopRequireDefault(require("color"));

var React = _interopRequireWildcard(require("react"));

var _Text = _interopRequireDefault(require("../Text"));

var _Icon = require("../Icon");

var _TouchableRipple = _interopRequireDefault(require("../TouchableRipple/TouchableRipple"));

var _styledComponents = require("styled-components");

var _native = _interopRequireDefault(require("styled-components/native"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _templateObject3() {
  const data = _taggedTemplateLiteral(["\n    margin-right: 32px;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  const data = _taggedTemplateLiteral(["\n    flex-direction: row;\n    align-items: center;\n    padding: 8px;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  const data = _taggedTemplateLiteral(["\n    margin-top: 10px;\n    margin-bottom: 10px;\n    margin-right: 4px;\n    margin-left: 4px;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const DrawerItemContainer = _native.default.View(_templateObject());

const DrawerItemWrapper = _native.default.View(_templateObject2());

const DrawerItemLabel = (0, _native.default)(_Text.default)(_templateObject3());

/**
 * A component used to show an action item with an icon and a label in a navigation drawer.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img class="medium" src="screenshots/drawer-item.png" />
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { DrawerItem } from 'react-native-simple-elements/components/Drawer';
 *
 * const MyComponent = () => (
 *    <Drawer.Item
 *      style={{ backgroundColor: '#64ffda' }}
 *      icon="star"
 *      label="First Item"
 *    />
 * );
 *
 * export default MyComponent;
 * ```
 */
const DrawerItem = ({
  icon,
  label,
  active,
  style,
  onPress,
  accessibilityLabel,
  ...rest
}) => {
  const theme = React.useContext(_styledComponents.ThemeContext);
  const {
    colors,
    roundness
  } = theme;
  const backgroundColor = active ? (0, _color.default)(colors.primary).alpha(0.12).rgb().string() : "transparent";
  const contentColor = active ? colors.primary : (0, _color.default)(colors.text).alpha(0.68).rgb().string();
  const font = theme.fonts.medium;
  const labelMargin = icon ? 32 : 0;
  return /*#__PURE__*/React.createElement(DrawerItemContainer, _extends({}, rest, {
    style: [{
      backgroundColor,
      borderRadius: roundness
    }, style]
  }), /*#__PURE__*/React.createElement(_TouchableRipple.default, {
    borderless: true // delayPressIn={0}
    ,
    onPress: onPress,
    style: {
      borderRadius: roundness
    } // accessibilityTraits={active ? ['button', 'selected'] : 'button'}
    // accessibilityComponentType="button"
    ,
    accessibilityRole: "button",
    accessibilityState: {
      selected: active
    },
    accessibilityLabel: accessibilityLabel
  }, /*#__PURE__*/React.createElement(DrawerItemWrapper, null, icon ? /*#__PURE__*/React.createElement(_Icon.SvgIcon, {
    icon: icon,
    size: 24,
    color: contentColor
  }) : null, /*#__PURE__*/React.createElement(DrawerItemLabel, {
    selectable: false,
    numberOfLines: 1,
    style: [{
      color: contentColor,
      ...font,
      marginLeft: labelMargin
    }]
  }, label))));
};

DrawerItem.displayName = "Drawer.Item";
var _default = DrawerItem;
exports.default = _default;
