// react-native-simple-elements/dist/commonjs/Shimmer/Shimmer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.createShimmerPlaceholder = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const getOutputRange = (width, isReversed) => isReversed ? [width, -width] : [-width, width];

const defaultProps = {
  delay: 0,
  duration: 1000,
  isInteraction: true
};

class ShimmerPlaceholder extends React.PureComponent {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "state", {
      beginShimmerPosition: new _reactNative.Animated.Value(-1)
    });

    _defineProperty(this, "getAnimated", () => {
      const {
        delay,
        duration,
        isInteraction
      } = this.props;
      return _reactNative.Animated.loop(_reactNative.Animated.timing(this.state.beginShimmerPosition, {
        toValue: 1,
        delay,
        duration,
        useNativeDriver: !["web"].includes(_reactNative.Platform.OS),
        isInteraction
      }));
    });

    _defineProperty(this, "animatedValue", this.getAnimated());
  }

  render() {
    return /*#__PURE__*/React.createElement(BasedShimmerPlaceholder, _extends({}, this.props, {
      animatedValue: this.animatedValue,
      beginShimmerPosition: this.state.beginShimmerPosition
    }));
  }

}

_defineProperty(ShimmerPlaceholder, "defaultProps", defaultProps);

const BasedShimmerPlaceholder = props => {
  const {
    width = 200,
    height = 15,
    // duration = 1000,
    // delay = 0,
    shimmerColors = ["#ebebeb", "#c5c5c5", "#ebebeb"],
    isReversed = false,
    stopAutoRun = false,
    visible,
    location = [0.3, 0.5, 0.7],
    style,
    contentStyle,
    shimmerStyle,
    // isInteraction = true,
    LinearGradient = _reactNative.View,
    // LinearGradient = global.Expo ? global.Expo.LinearGradient : View,
    children,
    animatedValue,
    beginShimmerPosition,
    shimmerWidthPercent = 1
  } = props;
  const linearTranslate = beginShimmerPosition.interpolate({
    inputRange: [-1, 1],
    outputRange: getOutputRange(width, isReversed)
  });
  React.useEffect(() => {
    if (!stopAutoRun) {
      animatedValue.start();
    }

    return () => {
      animatedValue.stop();
    };
  }, [stopAutoRun]);
  React.useEffect(() => {
    if (visible) {
      animatedValue.stop();
    }

    if (!visible && !stopAutoRun) {
      animatedValue.start();
    }
  }, [visible, stopAutoRun]);
  return /*#__PURE__*/React.createElement(_reactNative.View, {
    style: [!visible && {
      height,
      width
    }, styles.container, !visible && shimmerStyle, style]
  }, /*#__PURE__*/React.createElement(_reactNative.View, {
    style: [!visible && {
      width: 0,
      height: 0,
      opacity: 0
    }, visible && contentStyle]
  }, children), !visible && /*#__PURE__*/React.createElement(_reactNative.View, {
    style: {
      flex: 1,
      backgroundColor: shimmerColors[0]
    }
  }, /*#__PURE__*/React.createElement(_reactNative.Animated.View, {
    style: {
      flex: 1,
      transform: [{
        translateX: linearTranslate
      }]
    }
  }, /*#__PURE__*/React.createElement(LinearGradient, {
    colors: shimmerColors,
    style: {
      flex: 1,
      width: width * shimmerWidthPercent
    },
    start: {
      x: -1,
      y: 0.5
    },
    end: {
      x: 2,
      y: 0.5
    },
    locations: location
  }))));
};

const styles = _reactNative.StyleSheet.create({
  container: {
    overflow: "hidden"
  }
});
/**
 * To create ShimmerPlaceholder by Linear Gradient. Only useful when you use 3rd party,
 * For example: react-native-linear-gradient
 * @param {Linear Gradient Component} LinearGradient - 'expo-linear-gradient' by default
 *
 * @example
 *
 * import LinearGradient from 'react-native-linear-gradient';
 * import { createShimmerPlaceholder } from 'react-native-shimmer-placeholder'
 *
 * const ShimmerPlaceHolder = createShimmerPlaceholder(LinearGradient)
 *
 * ...
 *
 * <ShimmerPlaceHolder />
 */


const createShimmerPlaceholder = (LinearGradient = _reactNative.View) => /*#__PURE__*/React.forwardRef((props, ref) =>
/*#__PURE__*/
// @ts-ignore
React.createElement(ShimmerPlaceholder, _extends({
  LinearGradient: LinearGradient,
  ref: ref
}, props)));

exports.createShimmerPlaceholder = createShimmerPlaceholder;
var _default = ShimmerPlaceholder;
exports.default = _default;
