// react-native-simple-widgets/dist/commonjs/BigCalendar/hooks/useCalendarTouchableOpacityProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCalendarTouchableOpacityProps = useCalendarTouchableOpacityProps;

var _dayjs = _interopRequireDefault(require("dayjs"));

var _react = _interopRequireDefault(require("react"));

var _commonStyles = require("../styles/commonStyles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function useCalendarTouchableOpacityProps({
  event,
  eventCellStyle,
  injectedStyles = [],
  onPressEvent
}) {
  const getEventStyle = _react.default.useMemo(() => typeof eventCellStyle === "function" ? eventCellStyle : () => eventCellStyle, [eventCellStyle]);

  const plainJsEvent = _react.default.useMemo(() => ({ ...event,
    start: (0, _dayjs.default)(event.start).toDate(),
    end: (0, _dayjs.default)(event.end).toDate()
  }), [event]);

  const _onPress = _react.default.useCallback(() => {
    onPressEvent && onPressEvent(plainJsEvent);
  }, [onPressEvent, plainJsEvent]);

  const touchableOpacityProps = {
    delayPressIn: 20,
    key: event.start.toString(),
    style: [_commonStyles.eventCellCss.style, ...injectedStyles, getEventStyle(plainJsEvent)],
    onPress: _onPress,
    disabled: !onPressEvent
  };
  return touchableOpacityProps;
}
