// react-native-simple-widgets/dist/module/SwiperFlatList/Pagination/Pagination.js
import React from "react";
import { StyleSheet, TouchableOpacity, View } from "react-native";
import { colors, vertical, horizontal } from "../themes";
const styles = StyleSheet.create({
  container: {
    position: "absolute",
    flexDirection: "row",
    marginVertical: vertical.xxSmall,
    justifyContent: "center",
    alignSelf: "center",
    bottom: 0,
    height: horizontal.small
  },
  pagination: {
    width: horizontal.small,
    height: horizontal.small,
    borderRadius: 25,
    marginHorizontal: horizontal.xSmall
  }
});
export const Pagination = ({
  size,
  paginationIndex = 0,
  scrollToIndex,
  paginationDefaultColor = colors.gray,
  paginationActiveColor = colors.white,
  paginationStyle = {},
  paginationStyleItem = {},
  paginationStyleItemActive = {},
  paginationStyleItemInactive = {},
  onPaginationSelectedIndex,
  e2eID = ""
}) => {
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, paginationStyle]
  }, Array.from({
    length: size
  }).map((_, index) => /*#__PURE__*/React.createElement(TouchableOpacity, {
    testID: "".concat(e2eID, "_pagination_").concat(index),
    style: [styles.pagination, paginationStyleItem, paginationIndex === index ? {
      backgroundColor: paginationActiveColor
    } : {
      backgroundColor: paginationDefaultColor
    }, paginationIndex === index ? paginationStyleItemActive : paginationStyleItemInactive],
    key: index,
    onPress: () => {
      scrollToIndex({
        index
      });
      onPaginationSelectedIndex === null || onPaginationSelectedIndex === void 0 ? void 0 : onPaginationSelectedIndex();
    }
  })));
};
