// react-native-simple-elements/dist/commonjs/theme/DarkTheme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _color = _interopRequireDefault(require("color"));

var _LightTheme = _interopRequireDefault(require("./LightTheme"));

var _colors = require("./colors");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DarkTheme = { ..._LightTheme.default,
  dark: true,
  mode: "adaptive",
  colors: { ..._LightTheme.default.colors,
    primary: "#BB86FC",
    accent: "#03dac6",
    background: "#121212",
    surface: "#121212",
    error: "#CF6679",
    text: _colors.white,
    onBackground: "#FFFFFF",
    onSurface: "#FFFFFF",
    black: _colors.black,
    white: _colors.white,
    disabled: (0, _color.default)(_colors.white).alpha(0.38).rgb().string(),
    placeholder: (0, _color.default)(_colors.white).alpha(0.54).rgb().string(),
    backdrop: (0, _color.default)(_colors.black).alpha(0.5).rgb().string(),
    notification: _colors.pinkA100
  }
};
var _default = DarkTheme;
exports.default = _default;
