// react-native-simple-elements/dist/commonjs/theme/LightTheme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _color = _interopRequireDefault(require("color"));

var _colors = require("./colors");

var _fonts = _interopRequireDefault(require("./fonts"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const LightTheme = {
  dark: false,
  roundness: 4,
  colors: {
    primary: "#6200ee",
    // primary
    accent: "#03dac4",
    // secondary
    background: "#f6f6f6",
    // background
    surface: _colors.white,
    // surface
    error: "#B00020",
    // error
    text: _colors.black,
    // on primary
    onBackground: "#000000",
    // on background
    onSurface: "#000000",
    // on surface
    black: _colors.black,
    white: _colors.white,
    disabled: (0, _color.default)(_colors.black).alpha(0.26).rgb().string(),
    // disabled
    placeholder: (0, _color.default)(_colors.black).alpha(0.54).rgb().string(),
    // placeholder
    backdrop: (0, _color.default)(_colors.black).alpha(0.5).rgb().string(),
    // backdrop
    notification: _colors.pinkA400
  },
  fonts: (0, _fonts.default)(),
  animation: {
    scale: 1.0
  }
};
var _default = LightTheme;
exports.default = _default;
