// react-native-simple-elements/dist/module/RadioButton/RadioButtonItem.js
import * as React from "react";
import { View, StyleSheet } from "react-native";
import { RadioButtonContext } from "./RadioButtonGroup";
import { handlePress } from "./utils";
import TouchableRipple from "../TouchableRipple/TouchableRipple";
import Text from "../Text";
import RadioButtonAndroid from "./RadioButtonAndroid";
import RadioButtonIOS from "./RadioButtonIOS";
import { ThemeContext } from "styled-components";

/**
 * RadioButton.Item allows you to press the whole row (item) instead of only the RadioButton.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img class="medium" src="screenshots/radio-item.ios.png" />
 *     <figcaption>Pressed</figcaption>
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import RadioButton from 'react-native-simple-elements/components/RadioButton';
 *
 * const MyComponent = () => {
 *   const [value, setValue] = React.useState('first');
 *
 *   return (
 *     <RadioButton.Group onValueChange={value => setValue(value)} value={value}>
 *       <RadioButton.Item label="First item" value="first" />
 *       <RadioButton.Item label="Second item" value="second" />
 *     </RadioButton.Group>
 *   );
 * };
 *
 * export default MyComponent;
 *```
 */
const RadioButtonItem = ({
  value,
  label,
  style,
  labelStyle,
  onPress,
  disabled,
  color,
  uncheckedColor,
  status,
  accessibilityLabel,
  testID,
  mode,
  textAfterButton
}) => {
  const theme = React.useContext(ThemeContext);
  const {
    colors
  } = theme;
  const radioButtonProps = {
    value,
    disabled,
    status,
    color,
    uncheckedColor
  };
  let radioButton;

  if (mode === "android") {
    radioButton = /*#__PURE__*/React.createElement(RadioButtonAndroid, radioButtonProps);
  } else if (mode === "ios") {
    radioButton = /*#__PURE__*/React.createElement(RadioButtonIOS, radioButtonProps);
  } else {
    radioButton = /*#__PURE__*/React.createElement(RadioButtonAndroid, radioButtonProps);
  }

  return /*#__PURE__*/React.createElement(RadioButtonContext.Consumer, null, context => {
    return /*#__PURE__*/React.createElement(TouchableRipple, {
      onPress: disabled ? undefined : () => handlePress({
        onPress: onPress,
        onValueChange: context === null || context === void 0 ? void 0 : context.onValueChange,
        value
      }),
      accessibilityLabel: accessibilityLabel,
      testID: testID
    }, /*#__PURE__*/React.createElement(View, {
      style: [styles.container, style],
      pointerEvents: "none"
    }, !textAfterButton ? /*#__PURE__*/React.createElement(Text, {
      style: [styles.label, {
        color: colors.text
      }, labelStyle]
    }, label) : null, radioButton, textAfterButton ? /*#__PURE__*/React.createElement(Text, {
      style: [styles.label, {
        color: colors.text
      }, labelStyle]
    }, label) : null));
  });
};

RadioButtonItem.displayName = "RadioButton.Item";
export default RadioButtonItem; // @component-docs ignore-next-line

const RadioButtonItemWithTheme = RadioButtonItem; // @component-docs ignore-next-line

export { RadioButtonItemWithTheme as RadioButtonItem };
const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    paddingVertical: 8,
    paddingTop: 8,
    paddingBottom: 8,
    paddingHorizontal: 16,
    paddingRight: 16,
    paddingLeft: 16
  },
  label: {
    fontSize: 16,
    flex: 1
  }
});
