// react-native-simple-elements/dist/commonjs/Modal/Modalbk.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _reactNativeIphoneXHelper = require("react-native-iphone-x-helper");

var _styledComponents = require("styled-components");

var _Surface = _interopRequireDefault(require("../Surface"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const defaultProps = {
  dismissable: true,
  visible: false,
  overlayAccessibilityLabel: "Close modal"
};
const DEFAULT_DURATION = 220;
const TOP_INSET = (0, _reactNativeIphoneXHelper.getStatusBarHeight)(true);
const BOTTOM_INSET = (0, _reactNativeIphoneXHelper.getBottomSpace)();
/**
 * The Modal component is a simple way to present content above an enclosing view.
 * To render the `Modal` above other components, you'll need to wrap it with the [`Portal`](portal.html) component.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img class="medium" src="screenshots/modal.gif" />
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import Button from 'react-native-simple-elements/components/Button';
 * import Text from 'react-native-simple-elements/components/Text';
 * import Modal from 'react-native-simple-elements/components/Modal';
 * import Divider from 'react-native-simple-elements/components/Divider';
 * import Provider from 'react-native-simple-elements/components/theme/Provider';
 *
 * const MyComponent = () => {
 *   const [visible, setVisible] = React.useState(false);
 *
 *   const showModal = () => setVisible(true);
 *   const hideModal = () => setVisible(false);
 *   const containerStyle = {backgroundColor: 'white', padding: 20};
 *
 *   return (
 *     <Provider>
 *       <Portal>
 *         <Modal visible={visible} onDismiss={hideModal} contentContainerStyle={containerStyle}>
 *           <Text>Example Modal.  Click outside this area to dismiss.</Text>
 *         </Modal>
 *       </Portal>
 *       <Button style={{marginTop: 30}} onPress={showModal}>
 *         Show
 *       </Button>
 *     </Provider>
 *   );
 * };
 *
 * export default MyComponent;
 * ```
 */

const Modal = ({
  visible,
  dismissable,
  onDismiss,
  children,
  style,
  contentContainerStyle,
  overlayAccessibilityLabel
}) => {
  const [rendered, setRendered] = React.useState(visible);
  const [opacity] = React.useState(new _reactNative.Animated.Value(visible ? 1 : 0));
  const theme = React.useContext(_styledComponents.ThemeContext);
  const {
    colors
  } = theme;
  React.useEffect(() => {
    if (visible) {
      setRendered(true);
      showModal();
    } else {
      hideModal();
    }

    return () => {
      _reactNative.BackHandler.removeEventListener("hardwareBackPress", handleBack);
    };
  }, [visible]);

  const handleBack = () => {
    if (dismissable) {
      hideModal();
    }

    return true;
  };

  const showModal = () => {
    _reactNative.BackHandler.removeEventListener("hardwareBackPress", handleBack);

    _reactNative.BackHandler.addEventListener("hardwareBackPress", handleBack);

    const {
      scale
    } = theme.animation;

    _reactNative.Animated.timing(opacity, {
      toValue: 1,
      duration: scale * DEFAULT_DURATION,
      easing: _reactNative.Easing.out(_reactNative.Easing.cubic),
      useNativeDriver: true
    }).start();
  };

  const hideModal = () => {
    _reactNative.BackHandler.removeEventListener("hardwareBackPress", handleBack);

    const {
      scale
    } = theme.animation;

    _reactNative.Animated.timing(opacity, {
      toValue: 0,
      duration: scale * DEFAULT_DURATION,
      easing: _reactNative.Easing.out(_reactNative.Easing.cubic),
      useNativeDriver: true
    }).start(({
      finished
    }) => {
      if (!finished) {
        return;
      }

      if (visible && onDismiss) {
        onDismiss();
      }

      if (visible) {
        showModal();
      } else {
        setRendered(false);
      }
    });
  };

  if (!rendered) return null;
  return /*#__PURE__*/React.createElement(_reactNative.Animated.View, {
    pointerEvents: visible ? "auto" : "none",
    accessibilityViewIsModal: true,
    accessibilityLiveRegion: "polite",
    style: _reactNative.StyleSheet.absoluteFill,
    onAccessibilityEscape: hideModal
  }, /*#__PURE__*/React.createElement(_reactNative.TouchableWithoutFeedback, {
    accessibilityLabel: overlayAccessibilityLabel,
    accessibilityRole: "button",
    disabled: !dismissable,
    onPress: dismissable ? hideModal : undefined
  }, /*#__PURE__*/React.createElement(_reactNative.Animated.View, {
    style: [styles.backdrop, {
      backgroundColor: colors.backdrop,
      opacity
    }]
  })), /*#__PURE__*/React.createElement(_reactNative.View, {
    style: [styles.wrapper, {
      marginTop: TOP_INSET,
      marginBottom: BOTTOM_INSET
    }, style],
    pointerEvents: "box-none"
  }, /*#__PURE__*/React.createElement(_Surface.default, {
    style: [{
      opacity
    }, styles.content, contentContainerStyle]
  }, children)));
};

Modal.defaultProps = defaultProps;
var _default = Modal;
exports.default = _default;

const styles = _reactNative.StyleSheet.create({
  backdrop: {
    flex: 1
  },
  wrapper: { ..._reactNative.StyleSheet.absoluteFillObject,
    justifyContent: "center"
  },
  content: {
    backgroundColor: "transparent",
    justifyContent: "center"
  }
});
