// react-native-simple-widgets/dist/commonjs/SelectPicker/mweb/Popup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _PopupMixin = _interopRequireDefault(require("./PopupMixin"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _templateObject() {
  const data = _taggedTemplateLiteral(["\n    background-image: -webkit-linear-gradient(top, #e7e7e7, #e7e7e7, transparent, transparent);\n    background-image: linear-gradient(to bottom, #e7e7e7, #e7e7e7, transparent, transparent);\n    background-position: bottom;\n    background-size: 100% 1px;\n    background-repeat: no-repeat;\n    position: relative;\n    display: flex;\n    flex-direction: row;\n    &-left, &-right {\n        padding-left: 15px;\n        padding-right: 15px;\n    }\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const SelectPickerPopupHeader = _styledComponents.default.div(_templateObject());

const getModal = (props, visible, {
  getContent,
  hide,
  onDismiss,
  onOk
}) => {
  if (!visible) {
    return null;
  }

  const {
    prefixCls
  } = props;
  return /*#__PURE__*/React.createElement(_reactNative.Modal // prefixCls={`${prefixCls}`}
  // className={props.className || ''}
  , {
    visible: true // closable={false}
    // transitionName={props.transitionName || props.popupTransitionName}
    // maskTransitionName={props.maskTransitionName}
    // onClose={hide}
    ,
    style: props.style
  }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(SelectPickerPopupHeader, {
    className: "".concat(prefixCls, "-header")
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-item ").concat(prefixCls, "-header-left"),
    onClick: onDismiss
  }, props.dismissText), /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-item ").concat(prefixCls, "-title")
  }, props.title), /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-item ").concat(prefixCls, "-header-right"),
    onClick: onOk
  }, props.okText)), getContent()));
};

var _default = (0, _PopupMixin.default)(getModal, {
  prefixCls: "rmc-picker-popup",
  WrapComponent: "span",
  triggerType: "onClick",
  pickerValueProp: "selectedValue",
  pickerValueChangeProp: "onValueChange"
});

exports.default = _default;
