// react-native-simple-widgets/dist/module/BigCalendar/CalendarHeaderForMonthView.js
import dayjs from "dayjs";
import * as React from "react";
import { Text, View } from "react-native";
import { u } from "./styles/commonStyles";
import { useTheme } from "./theme/ThemeContext";
import { getDatesInWeek, typedMemo } from "./utils";

function _CalendarHeaderForMonthView({
  locale,
  weekStartsOn,
  style
}) {
  const dates = getDatesInWeek(new Date(), weekStartsOn, locale);
  const todayWeekNum = dayjs().day();
  const theme = useTheme();
  return /*#__PURE__*/React.createElement(View, {
    style: [u["border-b"], {
      borderColor: theme.palette.gray["100"]
    }, theme.isRTL ? u["flex-row-reverse"] : u["flex-row"], style]
  }, dates.map(date => /*#__PURE__*/React.createElement(View, {
    style: {
      flex: 1,
      paddingTop: 2
    },
    key: date.toISOString()
  }, /*#__PURE__*/React.createElement(View, {
    style: {
      height: 30
    }
  }, /*#__PURE__*/React.createElement(Text, {
    style: [u["text-center"], {
      color: todayWeekNum === date.day() ? theme.palette.primary.main : theme.palette.gray["800"]
    }]
  }, date.format("ddd"))))));
}

export const CalendarHeaderForMonthView = typedMemo(_CalendarHeaderForMonthView);
