// react-native-simple-elements/dist/commonjs/BottomSheet/BottomSheetWithRef.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _BottomSheet = _interopRequireDefault(require("./BottomSheet"));

var _Surface = _interopRequireDefault(require("../Surface"));

var _Container = require("../Container");

var _close = _interopRequireDefault(require("@mdi/svg/svg/close.svg"));

var _IconButton = _interopRequireDefault(require("../IconButton"));

var _Button = _interopRequireDefault(require("../Button"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const defaultProps = {
  translations: {
    confirmText: "Done"
  }
};

const BottomSheetWithRef = ({
  innerRef,
  children,
  title,
  translations,
  ...rest
}) => {
  const [visible, setVisible] = React.useState(false);
  const optionsRef = React.useRef({});
  React.useImperativeHandle(innerRef, () => ({
    open: options => {
      optionsRef.current = {
        onConfirm: options === null || options === void 0 ? void 0 : options.onConfirm,
        onCancel: options === null || options === void 0 ? void 0 : options.onCancel
      };
      setVisible(true);
    },
    close: () => {
      setVisible(false);
    }
  }));

  const _handleConfirm = () => {
    if (optionsRef && optionsRef.current && optionsRef.current.onConfirm) {
      optionsRef.current.onConfirm();
    }

    innerRef.current.close();
  };

  const _handleDismiss = () => {
    innerRef.current.close();
  };

  return /*#__PURE__*/React.createElement(_BottomSheet.default, _extends({
    visible: visible,
    onDismiss: _handleDismiss
  }, rest), /*#__PURE__*/React.createElement(_Surface.default, null, /*#__PURE__*/React.createElement(_Container.FlexItem, {
    title: title,
    left: () => /*#__PURE__*/React.createElement(_IconButton.default, {
      icon: _close.default,
      onPress: _handleDismiss
    }),
    right: () => /*#__PURE__*/React.createElement(_Button.default, {
      onPress: _handleConfirm
    }, translations.confirmText),
    rowStyle: {
      alignItems: "center"
    }
  }), children));
};

BottomSheetWithRef.defaultProps = defaultProps;
var _default = BottomSheetWithRef;
exports.default = _default;
