// react-native-simple-elements/dist/commonjs/AlertDialog/AlertDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _styledComponents = require("styled-components");

var _Surface = _interopRequireDefault(require("../Surface"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const WindowHeight = _reactNative.Dimensions.get("window").height;

const SUPPORTED_ORIENTATIONS = ["portrait", "portrait-upside-down", "landscape", "landscape-left", "landscape-right"];
// type State = {
//     animatedHeight: Animated.Value,
//     pan: Animated.ValueXY,
//     modalVisible: boolean,
// }
const defaultProps = {
  animationType: "none",
  openDuration: 100,
  closeDuration: 100,
  closeOnDragDown: false,
  dragFromTopOnly: false,
  closeOnPressBack: true,
  keyboardAvoidingViewEnabled: _reactNative.Platform.OS === "ios",
  customStyles: {},
  onDismiss: null,
  onOpen: null,
  children: /*#__PURE__*/React.createElement(_reactNative.View, null),
  backdropColor: "#00000077"
};

const AlertDialog = ({
  visible,
  dismissable,
  animationType,
  closeOnDragDown,
  dragFromTopOnly,
  closeOnPressBack,
  children,
  customStyles,
  keyboardAvoidingViewEnabled,
  openDuration,
  closeDuration,
  onOpen,
  onDismiss,
  backdropColor,
  contentContainerStyle,
  contentStyle
}) => {
  var _panResponder$current, _panResponder$current2;

  const [isVisible, setIsVisible] = React.useState(false);
  const theme = React.useContext(_styledComponents.ThemeContext);
  const opacityValue = React.useRef(new _reactNative.Animated.Value(0)).current;
  const pan = React.useRef(new _reactNative.Animated.ValueXY());
  const panResponder = React.useRef(undefined);
  React.useEffect(() => {
    if (!isVisible && visible) {
      setModalVisible(true);
    } else if (isVisible && !visible) {
      setModalVisible(false);
    }
  }, [visible]); // state: State = {
  //     modalVisible: false,
  //     animatedHeight: new Animated.Value(0),
  //     pan: new Animated.ValueXY()
  // };
  // panResponder;
  // constructor(props) {
  //     super(props);
  //     this.createPanResponder(props);
  // }

  const setModalVisible = visible => {
    // const { height, minClosingHeight, openDuration, closeDuration, onDismiss, onOpen } = this.props;
    // const { animatedHeight, pan } = this.state;
    if (visible) {
      setIsVisible(visible);
      if (typeof onOpen === "function") onOpen();

      _reactNative.Animated.timing(opacityValue, {
        toValue: 1,
        duration: openDuration || 100,
        useNativeDriver: true
      }).start();
    } else {
      _reactNative.Animated.timing(opacityValue, {
        toValue: 0,
        duration: closeDuration || 100,
        useNativeDriver: true
      }).start(() => {
        pan.current.setValue({
          x: 0,
          y: 0
        });
        setIsVisible(visible); // this.setState({
        //     modalVisible: visible,
        //     animatedHeight: new Animated.Value(0)
        // });

        if (typeof onDismiss === "function") onDismiss();
      });
    }
  };

  const createPanResponder = () => {
    // const { closeOnDragDown, height } = props;
    // const { pan } = this.state;
    panResponder.current = _reactNative.PanResponder.create({
      onStartShouldSetPanResponder: () => closeOnDragDown,
      onPanResponderMove: (e, gestureState) => {
        if (gestureState.dy > 0) {
          _reactNative.Animated.event([null, {
            dy: pan.current.y
          }], {
            useNativeDriver: false
          })(e, gestureState);
        }
      },
      onPanResponderRelease: (e, gestureState) => {
        // if (height / 4 - gestureState.dy < 0) {
        if (gestureState.dy > 15) {
          setModalVisible(false);
        } else {
          _reactNative.Animated.spring(pan.current, {
            toValue: {
              x: 0,
              y: 0
            },
            useNativeDriver: false
          }).start();
        }
      }
    });
  }; // const open = (props?) => {
  //     setModalVisible(true, props);
  // }


  const close = () => {
    setModalVisible(false);
  }; // const { animatedHeight, pan } = this.state;


  if (!visible && !panResponder.current) {
    createPanResponder();
  }

  const panStyle = {
    transform: pan.current.getTranslateTransform()
  };
  return /*#__PURE__*/React.createElement(_reactNative.Modal, {
    transparent: true,
    animationType: animationType,
    visible: isVisible,
    dismissable: dismissable // @ts-ignore
    ,
    supportedOrientations: SUPPORTED_ORIENTATIONS,
    onRequestClose: () => {
      if (closeOnPressBack) setModalVisible(false);
    }
  }, /*#__PURE__*/React.createElement(_reactNative.KeyboardAvoidingView, {
    enabled: keyboardAvoidingViewEnabled,
    behavior: "padding",
    style: [{
      flex: 1,
      backgroundColor: backdropColor
    }, customStyles.wrapper]
  }, /*#__PURE__*/React.createElement(_reactNative.TouchableWithoutFeedback // accessibilityLabel={overlayAccessibilityLabel}
  , {
    accessibilityRole: "button",
    disabled: !dismissable,
    onPress: dismissable ? close : undefined
  }, /*#__PURE__*/React.createElement(_reactNative.Animated.View, {
    style: [styles.backdrop, {
      flex: 1
    }, {
      backgroundColor: theme.colors.backdrop,
      opacity: opacityValue
    }]
  })), /*#__PURE__*/React.createElement(_reactNative.Animated.View, _extends({}, !dragFromTopOnly && (_panResponder$current = panResponder.current) !== null && _panResponder$current !== void 0 && _panResponder$current.panHandlers ? panResponder.current.panHandlers : {}, {
    style: [panStyle, {
      backgroundColor: theme.colors.backdrop,
      width: "100%",
      height: WindowHeight,
      overflow: "hidden"
    }, {
      flexDirection: "column",
      flex: 1,
      position: "absolute",
      top: 0,
      bottom: 0,
      left: 0,
      right: 0
    }, styles.contentContainer, contentContainerStyle, customStyles.container],
    pointerEvents: "box-none"
  }), closeOnDragDown && /*#__PURE__*/React.createElement(_reactNative.View, _extends({}, dragFromTopOnly && (_panResponder$current2 = panResponder.current) !== null && _panResponder$current2 !== void 0 && _panResponder$current2.panHandlers ? panResponder.current.panHandlers : {}, {
    style: {
      width: "100%",
      alignItems: "center",
      backgroundColor: "transparent"
    }
  }), /*#__PURE__*/React.createElement(_reactNative.View, {
    style: [{
      width: 35,
      height: 5,
      borderRadius: 5,
      margin: 10,
      backgroundColor: "#ccc"
    }, customStyles.draggableIcon]
  })), /*#__PURE__*/React.createElement(_reactNative.View, {
    style: {
      flex: 1
    },
    pointerEvents: "none"
  }), /*#__PURE__*/React.createElement(_Surface.default, {
    style: [styles.content, contentStyle]
  }, children), /*#__PURE__*/React.createElement(_reactNative.View, {
    style: {
      flex: 1
    },
    pointerEvents: "none"
  }))));
};

AlertDialog.defaultProps = defaultProps;
var _default = AlertDialog;
exports.default = _default;

const styles = _reactNative.StyleSheet.create({
  backdrop: {},
  contentContainer: {
    alignItems: "center"
  },
  content: {
    marginHorizontal: 16
  }
});
