// react-native-simple-widgets/dist/module/MultipleSlider/DefaultLabel.js
import * as React from "react";
import { View, Text, StyleSheet } from "react-native"; // const sliderRadius = 3;

const width = 50;
export default class DefaultLabel extends React.Component {
  render() {
    const {
      oneMarkerValue,
      twoMarkerValue,
      oneMarkerLeftPosition,
      twoMarkerLeftPosition,
      oneMarkerPressed,
      twoMarkerPressed
    } = this.props;
    return /*#__PURE__*/React.createElement(View, {
      style: {
        position: "relative"
      }
    }, Number.isFinite(oneMarkerLeftPosition) && Number.isFinite(oneMarkerValue) && /*#__PURE__*/React.createElement(View, {
      style: [styles.sliderLabel, {
        left: oneMarkerLeftPosition - width / 2
      }, oneMarkerPressed && styles.markerPressed]
    }, /*#__PURE__*/React.createElement(Text, {
      style: styles.sliderLabelText
    }, oneMarkerValue)), Number.isFinite(twoMarkerLeftPosition) && Number.isFinite(twoMarkerValue) && /*#__PURE__*/React.createElement(View, {
      style: [styles.sliderLabel, {
        left: twoMarkerLeftPosition - width / 2
      }, twoMarkerPressed && styles.markerPressed]
    }, /*#__PURE__*/React.createElement(Text, {
      style: styles.sliderLabelText
    }, twoMarkerValue)));
  }

}
const styles = StyleSheet.create({
  sliderLabel: {
    position: "absolute",
    bottom: 0,
    minWidth: width,
    padding: 8,
    backgroundColor: "#f1f1f1"
  },
  sliderLabelText: {
    alignItems: "center",
    textAlign: "center",
    fontStyle: "normal",
    fontSize: 11
  },
  markerPressed: {
    borderWidth: 2,
    borderColor: "#999"
  }
});
